src/wiki/plugins/links/mdx/urlize.py

```diff
@@ -2,98 +2,138 @@
-'<p><a href="http://example.com">example.com</a></p>'
-
->>> md.convert('example.net')
-'<p><a href="http://example.net">example.net</a></p>'
-
->>> md.convert('www.example.us')
-'<p><a href="http://www.example.us">www.example.us</a></p>'
-
->>> md.convert('(www.example.us/path/?name=val)')
-'<p>(<a href="http://www.example.us/path/?name=val">www.example.us/path/?name=val</a>)</p>'
+# Regular expression is meant to match the following pattern:
+#
+# [BEGIN_URL][PROTOCOL]HOST[:PORT][/[PATH]][END_URL]
+#
+# Everything except HOST is meant to be optional, as denoted by square
+# brackets.
+#
+# Patter elements are as follows:
+#
+# BEGIN_URL
+#   Either '<' or '('.
+#
+# PROTOCOL
+#   One of: 'http://', 'https://', 'ftp://', or 'ftps://'.
+#
+# HOST
+#   Host can be one of: IPv4 address, IPv6 address in full form, IPv6
+#   address in shortened form (e.g. ::1 vs 0:....:0:1 or any
+#   combination of), FQDN-like entry (dot-separated domain
+#   components), or string 'localhost'.
+#
+# PORT
+#   Port should be a numeric value. Keep in mind that it must be
+#   preceded with the colon (':').
+#
+# PATH
+#   Additional PATH, including any GET parameters that should be part
+#   of the URL.
+#
+# END_URL
+#   Either '>' or ')'. Should match with same type as BEGIN_URL.
+#
+# It should be noted that there are some inconsitencies with the below
+# regex, mainly that:
+#
+# - No IPv4 or IPv6 address validation is performed.
+# - Excessively long IPv6 addresses will end-up being matched if the
+#   shortened form happens somewhere in the middle of host string.
+#
+# In order to make the regex easier to handle later on, the following
+# named groups are provided:
+#
+# - begin (character denoting beginning of URL)
+# - url (entire URL that can be used, for example, as actual link for
+#   href).
+# - protocol (protocol, together with the trailing ://)
+# - host (just the host part)
+# - port (just the port number)
+# - path (path, combined with any additional GET parameters)
+# - end (character denoting end of URL)
+#
+URLIZE_RE = (
+    r'(?P<begin>[\(\<])?(?P<url>'  # begin url group
 
->>> md.convert('go to <http://example.com> now!')
-'<p>go to <a href="http://example.com">http://example.com</a> now!</p>'
+    # Leading protocol specification.
+    r'(?P<protocol>http://|https://|ftp://|ftps://|)'
 
-Negative examples:
+    # Host identifier
+    r'(?P<host>'  # begin host identifier group
 
->>> md.convert('del.icio.us')
-'<p>del.icio.us</p>'
+    r'[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}|'  # IPv4, match before FQDN
+    r'\[?([A-F0-9]{1,4}:){7}([A-F0-9]{1,4})\]?|'  # IPv6, full form
+    r'\[?:(:[A-F0-9]{1,4}){1,6}\]?|'  # IPv6, leading zeros removed
+    r'([A-F0-9]{1,4}:){1,6}:([A-F0-9]{1,4}){1,6}|'  # IPv6, zeros in middle removed.
+    r'\[?([A-F0-9]{1,4}:){1,6}:\]?|'  # IPv6, trailing zeros removed
+    r'\[?::\]?|'  # IPv6, just "empty" address
+    r'([A-Z0-9]([A-Z0-9-]{0,61}[A-Z0-9])?\.)+([A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'  # FQDN
```

Comment:
Great! We retain our `.horse` TLD support :)

Reply:
I actually had to go ahead and verify there _is_ a `.horse` TLD. And congratulations - I'm adding a test for it.